File: DigitalClock/Font.m
//
//  Font.m
//  DigitClock
//
//  Created by wenyou on 2017/7/10.
//  Copyright © 2017年 wenyou. All rights reserved.
//

#import "Font.h"

#import <CoreText/CoreText.h>
#import <AppKit/AppKit.h>


@implementation Font
static Font *_instance;

- (instancetype)init {
    if (self = [super init]) {
        NSDictionary<NSString *, NSString *> *fontNames = @{@"DigitalDismay": @"otf", @"DS-Digital": @"ttf"};
        [fontNames.allKeys enumerateObjectsUsingBlock:^(NSString * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            [self registFont:obj type:fontNames[obj]];
        }];
    }
    return self;
}

- (void)registFont:(NSString *)name type:(NSString *)type {
    NSString *path = [[NSBundle bundleForClass:Font.class] pathForResource:name ofType:type];
    NSData *dynamicFontData = [NSData dataWithContentsOfFile:path];
    if (!dynamicFontData) {
        return;
    }
    CGDataProviderRef providerRef = CGDataProviderCreateWithCFData((__bridge CFDataRef)dynamicFontData);
    CGFontRef font = CGFontCreateWithDataProvider(providerRef);
    CFErrorRef error;
    if (!CTFontManagerRegisterGraphicsFont(font, &error)) {
        //注册失败
        CFStringRef errorDescription = CFErrorCopyDescription(error);
        NSLog(@"Failed to load font: %@", errorDescription);
        CFRelease(errorDescription);
    }
    CFRelease(font);
    CFRelease(providerRef);
}

+ (instancetype)shareInstance {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [[self alloc] init];
    });
    return _instance ;
}

- (NSFont *)fontOfSize:(CGFloat)fontSize name:(NSString *)fontName {
    NSFont *font = [NSFont fontWithName:fontName size:fontSize];
    NSAssert(font != nil, @"%@ couldn't be loaded", fontName);
    return font;
}
@end

File: DigitalClock/Font.swift
//
//  Font.swift
//  DigitalClock
//
//  Created by wenyou on 2017/7/11.
//  Copyright © 2017年 wenyou. All rights reserved.
//

import AppKit

class Font {
    static let sharedInstance = Font()

    private init() {
        ["DigitalDismay": "otf", "DS-Digital": "ttf"].forEach { (key: String, value: String) in
            registFont(name: key, type: value)
        }
    }

    private func registFont(name: String, type: String) {
        if let path = Bundle.main.path(forResource: name, ofType: type), let dynamicFontData = NSData(contentsOfFile: path), let dataProvider = CGDataProvider(data: dynamicFontData) {
            let font = CGFont.init(dataProvider)
            var error: Unmanaged<CFError>? = nil
            if !CTFontManagerRegisterGraphicsFont(font, &error) {
                let errorDescription: CFString = CFErrorCopyDescription(error!.takeUnretainedValue())
                NSLog("Failed to load font: %@", errorDescription as String)
            }
            error?.release()
        }
    }

    func fontOfSize(fontSize: CGFloat, fontName: String) -> NSFont {
        if let font = NSFont.init(name: fontName, size: fontSize) {
            return font
        }
        return NSFont.systemFont(ofSize: fontSize)
    }
}
